### PR TITLE
Added codegen for destructors

### LINKED ISSUE #529: No bindings are being generated for destructors
This is probably mentioned elsewhere since it is such an obvious thing but I couldn't find a tracking issue for it so I thought I would create one.

### Input C/C++ Header

```C++
class Foo {
public:
    ~Foo();
};

Foo::~Foo() {
}
```

### Bindgen Invokation

```
$ bindgen test.hpp
```

### Actual Results

```rust
/* automatically generated by rust-bindgen */

#[repr(C)]
#[derive(Debug)]
pub struct Foo {
    pub _address: u8,
}
#[test]
fn bindgen_test_layout_Foo() {
    assert_eq!(::std::mem::size_of::<Foo>() , 1usize , concat ! (
               "Size of: " , stringify ! ( Foo ) ));
    assert_eq! (::std::mem::align_of::<Foo>() , 1usize , concat ! (
                "Alignment of " , stringify ! ( Foo ) ));
}
```

### Expected Results

```rust
// ... as above ...

impl Drop for Foo {
    fn drop(&mut self) {
      // ... call mangled destructor
    }
}
```

### PR DESCRIPTION
**Pull request by cynicaldevil**
Fixes #529

cc @emilio

### CODE REVIEW

File: tests/expectations/tests/union_dtor.rs [hunk 1]
```diff
@@ -52,3 +52,13 @@ fn bindgen_test_layout_UnionWithDtor() {
                 "Alignment of field: " , stringify ! ( UnionWithDtor ) , "::"
                 , stringify ! ( mBar ) ));
 }
+extern "C" {
+    #[link_name = "_ZN13UnionWithDtorD1Ev"]
+    pub fn UnionWithDtor_~UnionWithDtor(this: *mut UnionWithDtor);
```
Comment (emilio) [inline tests/expectations/tests/union_dtor.rs]: I don't think this can possibly compile? We need to mangle this name.

Comment (cynicaldevil) [inline tests/expectations/tests/union_dtor.rs]: But isn't that a valid identifier? I think Rust accepts all valid Unicode strings....

**Comment by fitzgen** [comment]
@cynicaldevil how's this going? Anything we can do to help get this across the finish line? Questions we can answer?

**Comment by cynicaldevil** [comment]
@fitzgen I have my midterms right now, but finally got a break in between, so I can work on this now. Any suggestions on what the mangled name of the destructor should be?

**Comment by fitzgen** [comment]
> Any suggestions on what the mangled name of the destructor should be?

`_bindgen_destructor_<class-or-enum-name>` should be OK.

**Comment by cynicaldevil** [comment]
So I'm trying to name the function as `*class_name*__bindgen__destructor` so that it is consistent with how the constructor is named, and I'm using the `canonical_name` method to get the whole string, remove the part after the tilde and append `__bindgen_destructor__` to it. Is there a better way of doing this?

**Comment by fitzgen** [comment]
@cynicaldevil is the `canonical_name` returning a name with the `~` in it? That is not what I would expect.

The `Item::name` method returns a `NameOptions` which gives a little bit more control over the name.

I would have expected `format!("_bindgen_desctructor_{}", item.canonical_name())` to Just Work, and I'm a bit surprised it isn't...

**Comment by emilio** [comment]
It isn't because the name of the destructor starts with `~`. We can strip it out while parsing though.

**Comment by cynicaldevil** [comment]
How can we strip away the `~` during parsing? I can't find anywhere to do it except in codegen.

**Comment by fitzgen** [comment]
Presumably in the `Function`/`FunctionSig` parsing code in `src/ir/function.rs`. I'm not really convinced that's any better or worse than in codegen, however.

I suggest just getting it working and then we can look at the code in review and decide if we think there is a better way or not.

**Comment by cynicaldevil** [comment]
@fitzgen After adding countless `debug!` statements to understand the flow of execution, I managed to get it to work as intended :)

**Comment by fitzgen** [comment]
@cynicaldevil It looks like this is failing a few tests with libclang 3.8 on Travis CI:

```diff
 extern "C" {
-    #[link_name = "_ZN7nsSlotsD1Ev"]
+    #[link_name = "_ZN7nsSlotsD0Ev"]
     pub fn nsSlots_nsSlots_destructor(this: *mut nsSlots);
 }
```

Do these test expectations just need to be updated or do we need to ignore these tests with libclang <= 3.8?

If the latter, you can add `// bindgen-unstable` to the top of the tests.

**Comment by cynicaldevil** [comment]
I don't know why this is happening, all my expectations are already up-to-date, and I am using clang 3.9.

**Comment by gsingh93** [comment]
Bump. I'm waiting for this feature :)

**Comment by emilio** [comment]
Ok, I just looked into why tests are failing. That test has a virtual destructor, that we should avoid generating. So I'm going to pull this and land it with that fix.

**Comment by emilio** [comment]
(I mean, it's not technically wrong to generate it, but I'd rather just do the proper thing for virtual methods instead)

**Comment by emilio** [comment]
Superseded by #608, thanks for this work @cynicaldevil! :)

**Pull request closed without merging**